FixMath.h

```diff
@@ -0,0 +1,1047 @@
+    typedef typename IntegerType<UBITSTOBYTES(new_NI+new_NF)>::unsigned_type return_type;
+    UFixMath<new_NI,new_NF> left(*this);
+    UFixMath<new_NI,new_NF> right(op);
+
+    return_type tt = return_type(left.asRaw()) + right.asRaw();
+    return UFixMath<new_NI,new_NF>(tt,true);
+  }
+
+  /** Addition with another type. Unsafe
+      @param op The number to be added.
+      @return The result of the addition as a UFixMath.
+  */
+  template<typename T>
+  UFixMath<NI,NF> operator+ (const T op) const
+  {
+    return UFixMath<NI,NF>(internal_value+((internal_type)op<<NF),true);
+  }
+
+
+  //////// SUBSTRACTION OVERLOADS
+
+  /** Subtraction with another UFixMath. Safe.
+      @param op The UFixMath to be subtracted.
+      @return The result of the subtraction as a SFixMath.
+  */
+  template<byte _NI, byte _NF> // We do not have the +1 after MAX(NI, _NI) because the substraction between two UFix should fit in the biggest of the two.
+  SFixMath<MAX(NI,_NI),MAX(NF,_NF)> operator- (const UFixMath<_NI,_NF>& op) const
+  {
+    constexpr byte new_NI = MAX(NI, _NI);
+    constexpr byte new_NF = MAX(NF, _NF);
+    typedef typename IntegerType<SBITSTOBYTES(new_NI+new_NF)>::signed_type return_type;
+    SFixMath<new_NI,new_NF> left(*this);
+    SFixMath<new_NI,new_NF> right(op);
+
+    return_type tt = return_type(left.asRaw()) - right.asRaw();
+    return SFixMath<new_NI,new_NF>(tt,true);
+  }
+
+  
+  /** Subtraction with another type. Unsafe
+      @param op The number to be subtracted.
+      @return The result of the subtraction as a UFixMath.
+  */
+  template<typename T>
+  UFixMath<NI,NF> operator- (const T op) const
+  {
+    return UFixMath<NI,NF>(internal_value-((internal_type)op<<NF),true);
+  }
+
+  /** Opposite of the number.
+      @return The opposite numberas a SFixMath.
+  */
+  SFixMath<NI,NF> operator-() const
+  {
+    return SFixMath<NI,NF>( -(typename IntegerType<SBITSTOBYTES(NI+NF)>::signed_type)(internal_value),true);
+  }
+
+  //////// MULTIPLICATION OVERLOADS
+  
+  /** Multiplication with another UFixMath. Safe.
+      @param op The UFixMath to be multiplied.
+      @return The result of the multiplication as a UFixMath.
+  */
+  template<byte _NI, byte _NF>
+  UFixMath<NI+_NI,NF+_NF> operator* (const UFixMath<_NI,_NF>& op) const
+  {
+    //typedef typename IntegerType< ((NI+_NI+NF+_NF-1)>>3)+1>::unsigned_type return_type ;
+    typedef typename IntegerType<UBITSTOBYTES(NI+_NI+NF+_NF)>::unsigned_type return_type ;
+    return_type tt = return_type(internal_value)*op.asRaw();
+    return UFixMath<NI+_NI,NF+_NF>(tt,true);
+  }
+
+  /** Multiplication with another type. Unsafe.
+      @param op The number to be multiplied.
+      @return The result of the multiplication as a UFixMath.
+  */
+  template<typename T>
+  UFixMath<NI,NF> operator* (const T op) const
+  {
+    return UFixMath<NI,NF>(internal_value*op,true);
+  }
+
+
+  //////// INVERSE
+  UFixMath<NF,NI> invFast() const
```

Comment:
These look really nifty, but I must admit, I don't understand them. Looking forward to your examples!

Reply:
Yeah, I thought having the inverse (normally computed in a slow loop, like `updateControl()` could be useful for having fast division, by multiplying by the inverse.

However, like most things I try to do, the great mind thought plan does not turn as expected (and I lose part of my night on it ;)).

Will do a clean description somewhere but here are starters, (which also help me to structure my mind, you don't have to read it all:

The basic idea, instead of dividing one by the number (which would always give zeros with integer) is to do:

((1<<(NI+NF))/a)>>(NI+NF).

Actually, in order to fit in the container, I use 1>>(NI+NF) - 1, but that does not change the matter much. Now for the output sizes:

A FixMath number is ruled by both its *range*, given by NI and its precision, given by NF. These are fundamentally different as increasing NF does not increase the range (which is is [0,1[) by NI does.

The inverse of any number `a`, actually has a reverse relationship: the bigger the initial range, the bigger the resolution of the inverse and vice-versa. This explains the implementation of `invFast()`. With this inverse, both the biggest number, and the smallest one are representable and have correct values: the inverse of the biggest one becomes the resolution and the inverse of the smallest one becomes the range. But there is a small problem: because 1/x is not linear (at all), this leads to a lot of duplicated values: if you take a UFixMath<8,0>, all values between 128 and 255 will have the same inverse, because the resolution is 1/255.

To solve that, one needs to compute the needed resolution so that the inverse of the *biggest number* and the *before biggest* do lead to different results, this is `invAccurate()`. Of course, the size completely explodes doing that so it should be used only when needed (or should be downcasted slightly after). I found the relation of the templates arguments by chance initially but I now have a more or less clean demonstration (actually the relation there is only valid to the first order, but that yields quite good results).

Now I understand why I saw written the division in fixed point arithmetic is tricky, and why floating point is better in that respect. People should have stumbled on the same problems… 40 years ago…